pkg/controllers/internalmembercluster/v1beta1/member_controller_integration_test.go

```diff
@@ -2,220 +2,834 @@
+	propertiesManuallyUpdatedConditionMsg1    = "Properties have been manually updated"
+	propertiesManuallyUpdatedConditionReason2 = "NewPropertiesPushedAgain"
+	propertiesManuallyUpdatedConditionMsg2    = "Properties have been manually updated again"
+)
 
-	AfterEach(func() {
-		By("delete member cluster namespace")
-		ns := corev1.Namespace{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: memberClusterNamespace,
-			},
-		}
-		Expect(k8sClient.Delete(ctx, &ns)).Should(Succeed())
-
-		By("delete member cluster nodes")
-		for _, node := range nodes.Items {
-			node := node
-			Expect(k8sClient.Delete(ctx, &node)).Should(Succeed())
-		}
-
-		By("delete member cluster")
-		internalMemberCluster := clusterv1beta1.InternalMemberCluster{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      memberClusterName,
-				Namespace: memberClusterNamespace,
-			},
-		}
-		Expect(k8sClient.Delete(ctx, &internalMemberCluster)).Should(SatisfyAny(Succeed(), &utils.NotFoundMatcher{}))
-	})
+const (
+	eventuallyTimeout  = time.Second * 30
+	eventuallyInterval = time.Millisecond * 500
+)
+
+var _ = Describe("Test InternalMemberCluster Controller", func() {
+	// Note that specs in this context run in serial, however, they might run in parallel with
+	// the other contexts if parallelization is enabled.
+	//
+	// This is safe as the controller managers have been configured to watch only their own
+	// respective namespaces.
+	Context("Test setup with property provider", Ordered, func() {
+		var (
+			// Add an offset of -1 second to avoid flakiness caused by approximation.
+			timeStarted = metav1.Time{Time: time.Now().Add(-time.Second)}
 
-	Context("join", func() {
-		BeforeEach(func() {
-			internalMemberCluster := clusterv1beta1.InternalMemberCluster{
+			// The timestamps below are set in later steps.
+			timeUpdated metav1.Time
+			timeLeft    metav1.Time
+		)
+
+		BeforeAll(func() {
+			// Create the InternalMemberCluster object.
+			imc := &clusterv1beta1.InternalMemberCluster{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:      memberClusterName,
-					Namespace: memberClusterNamespace,
+					Name:      member1Name,
+					Namespace: member1ReservedNSName,
 				},
 				Spec: clusterv1beta1.InternalMemberClusterSpec{
-					State:                  clusterv1beta1.ClusterStateJoin,
-					HeartbeatPeriodSeconds: int32(HBPeriod),
+					State: clusterv1beta1.ClusterStateJoin,
+					// Use a shorter heartbeat period to improve responsiveness.
+					HeartbeatPeriodSeconds: 2,
 				},
 			}
-			Expect(k8sClient.Create(ctx, &internalMemberCluster)).Should(Succeed())
-		})
+			Expect(hubClient.Create(ctx, imc)).Should(Succeed())
 
-		It("should update internalMemberCluster to joined", func() {
-			result, err := r.Reconcile(ctx, ctrl.Request{
-				NamespacedName: memberClusterNamespacedName,
+			// Report properties via the property provider.
+			observationTime := metav1.Now()
+			propertyProvider1.Update(&propertyprovider.PropertyCollectionResponse{
```

Comment:
IIRC, we setup the response here? If so, it's more like a UT then?

Reply:
Hi Ryan! This case is a bit different; it uses a provider and the data reported by the provider is manually refreshed. For the IMC controller integration test it might not matter much to actually implement a provider that watches for nodes/pods, as the behavior to verify only concerns how IMC controller processes data as reported by the provider.